File: android/app/src/main/java/com/nexusriot/rezoagwe/core/Runtime.kt
package com.nexusriot.rezoagwe.core

import android.content.Context
import com.nexusriot.rezoagwe.proto.Codec
import java.io.File
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** Everything the user can configure, persisted in shared preferences. */
data class Settings(
    val nick: String = "phone",
    val port: Int = 3137,
    val advertiseHost: String = "",
    val seeds: String = "",
    val psk: String = "",
    val cluster: String = Codec.DEFAULT_CLUSTER,
    val bootstrapPort: Int = 9999,
    val tombstoneTtlSec: Long = 0,
) {
    fun seedList(): List<String> =
        seeds.split(',').map { it.trim() }.filter { it.isNotEmpty() }

    /**
     * The values as they should be stored: surrounding whitespace off every text
     * field, and blanks replaced from [fallback].
     *
     * The cluster name and the key feed the framing key, so a stray space is not
     * cosmetic — it changes the key and the node stops talking to every peer,
     * with nothing on screen to show why. A soft keyboard supplies exactly that:
     * it appends a space after a word it thinks it has completed, so typing
     * "e2e" into the cluster field stores "e2e ". Trimming here rather than at
     * each call site is the point — the seeds and the advertised host were
     * already trimmed, and the two fields that mattered most were not.
     */
    fun sanitized(fallback: Settings = Settings()): Settings = copy(
        nick = nick.trim().ifBlank { fallback.nick },
        advertiseHost = advertiseHost.trim(),
        seeds = seeds.trim(),
        psk = psk.trim(),
        cluster = cluster.trim().ifBlank { fallback.cluster },
    )

    fun toNodeConfig() = NodeConfig(
        advertiseHost = advertiseHost,
        port = port,
        nick = nick,
        seeds = seedList(),
        psk = psk,
        cluster = cluster,
        tombstoneTtlSec = tombstoneTtlSec,
    )

    fun toBootstrapConfig() = BootstrapConfig(
        port = bootstrapPort,
        psk = psk,
        cluster = cluster,
    )
}

private const val PREFS = "rezoagwe"

object SettingsStore {
    fun load(context: Context): Settings {
        val p = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)
        val defaults = Settings()
        return Settings(
            nick = p.getString("nick", defaults.nick)!!,
            port = p.getInt("port", defaults.port),
            advertiseHost = p.getString("advertiseHost", defaults.advertiseHost)!!,
            seeds = p.getString("seeds", defaults.seeds)!!,
            psk = p.getString("psk", defaults.psk)!!,
            cluster = p.getString("cluster", defaults.cluster)!!,
            bootstrapPort = p.getInt("bootstrapPort", defaults.bootstrapPort),
            tombstoneTtlSec = p.getLong("tombstoneTtlSec", defaults.tombstoneTtlSec),
        )
    }

    fun save(context: Context, s: Settings) {
        context.getSharedPreferences(PREFS, Context.MODE_PRIVATE).edit()
            .putString("nick", s.nick)
            .putInt("port", s.port)
            .putString("advertiseHost", s.advertiseHost)
            .putString("seeds", s.seeds)
            .putString("psk", s.psk)
            .putString("cluster", s.cluster)
            .putInt("bootstrapPort", s.bootstrapPort)
            .putLong("tombstoneTtlSec", s.tombstoneTtlSec)
            .apply()
    }
}

/**
 * Owns the node and the rendezvous service for the whole process.
 *
 * Both roles are singletons rather than per-screen objects: they hold sockets and
 * survive the UI, which is the entire point of running them behind a foreground
 * service.
 */
object Runtime {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    // The engine is a flow, not a field: changing the port or the key rebuilds it,
    // and the UI has to follow the new instance rather than keep rendering a
    // stopped one.
    private val _engine = MutableStateFlow<NodeEngine?>(null)
    val engineFlow: StateFlow<NodeEngine?> = _engine.asStateFlow()

    private val _bootstrap = MutableStateFlow<BootstrapServer?>(null)
    val bootstrapFlow: StateFlow<BootstrapServer?> = _bootstrap.asStateFlow()

    private var engineRef: NodeEngine?
        get() = _engine.value
        set(value) {
            _engine.value = value
        }

    private var bootstrapRef: BootstrapServer?
        get() = _bootstrap.value
        set(value) {
            _bootstrap.value = value
        }

    private val _settings = MutableStateFlow(Settings())
    val settings: StateFlow<Settings> = _settings.asStateFlow()

    private val _lastError = MutableStateFlow<String?>(null)
    val lastError: StateFlow<String?> = _lastError.asStateFlow()

    fun init(context: Context) {
        if (engineRef != null) return
        _settings.value = SettingsStore.load(context)
        engineRef = NodeEngine(_settings.value.toNodeConfig(), stateFile(context), scope)
        bootstrapRef = BootstrapServer(_settings.value.toBootstrapConfig(), rosterFile(context), scope)
    }

    fun engine(context: Context): NodeEngine {
        init(context)
        return engineRef!!
    }

    fun bootstrap(context: Context): BootstrapServer {
        init(context)
        return bootstrapRef!!
    }

    /**
     * Applies new settings, rebuilding whichever role they affect.
     *
     * The port, cluster and key are baked into a running socket and codec, so a
     * change means a restart — done here rather than silently ignored, which is
     * the failure mode that makes settings screens untrustworthy.
     */
    fun applySettings(context: Context, updated: Settings) {
        init(context)
        val previous = _settings.value
        SettingsStore.save(context, updated)
        _settings.value = updated

        val engine = engineRef!!
        val nodeChanged = previous.port != updated.port ||
            previous.psk != updated.psk ||
            previous.cluster != updated.cluster ||
            previous.advertiseHost != updated.advertiseHost ||
            previous.seeds != updated.seeds ||
            previous.tombstoneTtlSec != updated.tombstoneTtlSec
        if (nodeChanged) {
            val wasRunning = engine.isRunning
            engine.stop()
            engineRef = NodeEngine(updated.toNodeConfig(), stateFile(context), scope)
            if (wasRunning) startNode(context)
        } else if (previous.nick != updated.nick) {
            engine.renameSelf(updated.nick)
        }

        val bootstrap = bootstrapRef!!
        if (previous.bootstrapPort != updated.bootstrapPort ||
            previous.psk != updated.psk ||
            previous.cluster != updated.cluster
        ) {
            val wasRunning = bootstrap.isRunning
            bootstrap.stop()
            bootstrap.config = updated.toBootstrapConfig()
            if (wasRunning) bootstrap.start()
        }
    }

    fun startNode(context: Context) = guard { engine(context).start() }

    fun stopNode(context: Context) = guard { engine(context).stop() }

    fun startBootstrap(context: Context) = guard { bootstrap(context).start() }

    fun stopBootstrap(context: Context) = guard { bootstrap(context).stop() }

    fun clearError() {
        _lastError.value = null
    }

    /** Binding a port is the one operation a user can plausibly get wrong; surface it instead of crashing. */
    private inline fun guard(block: () -> Unit) {
        try {
            block()
        } catch (e: Exception) {
            _lastError.value = e.message ?: e.toString()
        }
    }

    private fun stateFile(context: Context) = File(context.filesDir, "node.json")

    private fun rosterFile(context: Context) = File(context.filesDir, "bootstrap.json")
}

File: electron/src/core/runtime.js
'use strict';

const path = require('node:path');
const { EventEmitter } = require('node:events');

const { NodeEngine } = require('./node-engine');
const { BootstrapServer } = require('./bootstrap-server');
const {
  SettingsStore, toNodeConfig, toBootstrapConfig, nodeNeedsRestart, bootstrapNeedsRestart,
} = require('./settings');

/**
 * Owns both roles for the whole process.
 *
 * The engine is replaced rather than reconfigured when the port, cluster or key
 * changes: those are baked into a bound socket and a derived framing key, so
 * changing them means a restart. Doing it here rather than silently ignoring the
 * change is what keeps a settings screen trustworthy.
 *
 * It re-emits every engine event under a stable name, so the main process can
 * forward to the renderer without rewiring after a rebuild.
 */
class Runtime extends EventEmitter {
  constructor(dataDir, opts = {}) {
    super();
    this.dataDir = dataDir;
    this.opts = opts;
    this.settingsStore = new SettingsStore(path.join(dataDir, 'settings.json'));
    this.engine = null;
    this.bootstrap = null;
    this.lastError = null;
    this.build();
  }

  get settings() {
    return this.settingsStore.get();
  }

  build() {
    const s = this.settings;
    this.engine = new NodeEngine(toNodeConfig(s), path.join(this.dataDir, 'node.json'), this.opts);
    this.bootstrap = new BootstrapServer(
      toBootstrapConfig(s), path.join(this.dataDir, 'bootstrap.json'), this.opts,
    );
    this.wire();
  }

  wire() {
    for (const event of ['entries', 'peers', 'chat', 'activity', 'status', 'metrics', 'topology']) {
      this.engine.on(event, (payload) => this.emit(event, payload));
    }
    this.bootstrap.on('roster', (payload) => this.emit('bootstrap-roster', payload));
    this.bootstrap.on('status', (payload) => this.emit('bootstrap-status', payload));
    this.bootstrap.on('warning', (text) => this.emit('warning', text));
  }

  /**
   * Runs an operation that touches a socket, reporting the failure rather than
   * letting it escape. Binding a port is the one thing a user can plausibly get
   * wrong, and an unhandled rejection there would take the window with it.
   */
  async guard(fn) {
    try {
      await fn();
      return { ok: true };
    } catch (e) {
      this.lastError = e.message || String(e);
      this.emit('error-message', this.lastError);
      return { ok: false, error: this.lastError };
    }
  }

  startNode() {
    return this.guard(() => this.engine.start());
  }

  stopNode() {
    return this.guard(() => this.engine.stop());
  }

  startBootstrap() {
    return this.guard(() => this.bootstrap.start());
  }

  stopBootstrap() {
    return this.guard(() => this.bootstrap.stop());
  }

  /** Applies new settings, restarting whichever role they affect and preserving what was running. */
  async applySettings(partial) {
    const previous = this.settings;
    const updated = this.settingsStore.save(partial);

    if (nodeNeedsRestart(previous, updated)) {
      const wasRunning = this.engine.isRunning;
      await this.guard(() => this.engine.stop());
      this.engine.removeAllListeners();
      this.engine = new NodeEngine(toNodeConfig(updated), path.join(this.dataDir, 'node.json'), this.opts);
      this.bootstrap.removeAllListeners();
      this.wire();
      this.engine.publishEntries();
      this.engine.publishPeers();
      this.engine.publishChat();
      this.engine.publishMetrics();
      if (wasRunning) await this.startNode();
    } else if (previous.nick !== updated.nick) {
      await this.guard(() => this.engine.renameSelf(updated.nick));
    }

    if (bootstrapNeedsRestart(previous, updated)) {
      const wasRunning = this.bootstrap.isRunning;
      await this.guard(() => this.bootstrap.stop());
      this.bootstrap.config = { ...this.bootstrap.config, ...toBootstrapConfig(updated) };
      if (wasRunning) await this.startBootstrap();
      this.bootstrap.publish();
    }
    return updated;
  }

  /** The full current state, for a renderer that has just connected or reloaded. */
  snapshot() {
    return {
      settings: this.settings,
      status: this.engine.status(),
      entries: this.engine.entries(),
      peers: this.engine.peerList(),
      chat: this.engine.chat(),
      activity: this.engine.activity(),
      metrics: this.engine.metrics.snapshot(),
      topology: this.engine.topology(),
      bootstrapStatus: this.bootstrap.status(),
      bootstrapRoster: this.bootstrap.roster(),
    };
  }

  async shutdown() {
    await this.guard(() => this.engine.stop());
    await this.guard(() => this.bootstrap.stop());
  }
}

module.exports = { Runtime };
